### PR TITLE
Add new plugin discovery capabilities

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug
Half of the work for: https://github.com/NuGet/Home/issues/6704
Regression: No  
If Regression then when did it last work:   
If Regression then how are we preventing it in future:   

## Fix
Details: 

*Note* that this *targets* the *feature* branch. 

As per the spec  here: 
https://github.com/NuGet/Home/wiki/NuGet-cross-plat-authentication-plugin#plugin-installation-and-discovery

Copying the relevant information for simplicity below:

The plugins will be discovered as follows:

1. An environment variable NUGET_PLUGIN_PATHS, priority reserved of the plugins reserved. If the NUGET_PLUGIN_PATHS environment variable is set, it overrides the convention based plugin discovery.

    The environment variable should contain a full path to the executable, exe in the .NET Framework case and dll in the .NET Core case. It's at the user's discretion to make sure that the plugins are executable under that runtime.
2. User-location - The NuGet Home location - %UserProfile%/.nuget/plugins/

    .NET Core based plugins should be installed in:
    > %UserProfile%/.nuget/plugins/netcore

    .NET Framework based plugins should be installed in:
    > %UserProfile%/.nuget/plugins/netfx

    Each plugin will be installed in it's own folder.

    The plugin entry point will be the name of the installed folder, with the .dll extensions for .NET Core, and .exe extension for .NET Framework.

    Example:

    ```
    .nuget
        plugins
            netfx
                myFeedCredentialProvider
                    myFeedCredentialProvider.exe
                    nuget.protocol.dll
            netcore
                myFeedCredentialProvider
                    myFeedCredentialProvider.dll
                    nuget.protocol.dll
    ``` 

    There's a gap in this approach.

3. Predetermined location in Visual Studio and dotnet (relative to MsBuild.exe)

    The fixed location would be a folder such as, nugetplugins and will follow the same directory rules as listed under 2.

All plugins should be self-contained, and install all their dependencies in their respective folders.

Issues with the discovery of the user-location plugins

Different versions of nuget.exe/dotnet.exe could be using the same plugins.
The obvious issue arises when an older client tries to execute a newer plugin.
If a dotnet.exe under the 2.x runtime, tries to launch a 3.x plugin.
A potential approach would involve adding the target framework under which the plugin has been built.
As this a corner-case scenario that's easily resolvable, the current approach is that we go with the cleaner approach of not including any compatibility checks.

An additional sideeffect of discovering msbuild for every nuget operation is an extra message about the MSBuild discovery whenever nuget.exe is launched. 

The leftover work is actually coding in the built in path for both dotnet.exe and VS/MSBuild shipping vehicles. 
We could consider lowering the verbosity of this message. 

## Testing/Validation
Tests Added: Yes  
Reason for not adding tests:  
Validation done:

### CODE REVIEW

**Comment by nkolev92** [comment]
https://github.com/NuGet/NuGet.Client/pull/2208 fixes the msbuild discovery issue. It'll only be done as a last resort, whenever the client gets a 401.

**Comment by nkolev92** [comment]
🔔 @NuGet/nuget-client

File: src/NuGet.Clients/NuGet.CommandLine/Commands/Command.cs [hunk 1]
```diff
@@ -163,12 +179,19 @@ protected virtual bool ShouldOutputNuGetVersion
...
+        {
+            SetDefaultCredentialProvider(MsBuildDirectory);
+        }
+
         /// <summary>
         /// Set default credential provider for the HttpClient, which is used by V2 sources.
         /// Also set up authenticated proxy handling for V3 sources.
         /// </summary>
-        protected void SetDefaultCredentialProvider()
+        protected void SetDefaultCredentialProvider(Lazy<string> msbuildDirectory)
```
Comment (rohit21agrawal) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/Command.cs:191]: wouldn't it make more sense for the function signature to just accept a string, and the caller pass MsBuildDirectory.Value to the function call?

File: src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using NuGet.Common;
```
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs:4]: nit : sort usings

File: src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+namespace NuGet.Protocol.Plugins
+{
+    public static class PluginDiscoveryUtility
+    {
+        public static string InternalPluginDiscoveryRoot { get; set; } = null;
+
+        public static string GetInternalPlugins()
+        {
+            var rootDirectory = InternalPluginDiscoveryRoot ?? System.Reflection.Assembly.GetEntryAssembly()?.Location;;
```
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs]: while it's okay to be safe, will GetEntryAsssembly ever return null though? Being over safe sometimes masks errors

File: src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+                foreach (var pluginDirectory in pluginDirectories)
+                {
+#if IS_DESKTOP
+                    var expectedPluginName = Path.Combine(pluginDirectory, Path.GetFileName(pluginDirectory) + ".exe");
+#else
+                    var expectedPluginName = Path.Combine(pluginDirectory, Path.GetFileName(pluginDirectory) + ".dll");
+#endif
+
+                    var filesInDirectory = Directory.EnumerateFiles(pluginDirectory);
+                    if (filesInDirectory.Contains(expectedPluginName, PathUtility.GetStringComparerBasedOnOS()))
```
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs]: can't you just use File.Exists since you already have the full path? Why do you need to enumerate files?

**Review by rohit21agrawal — COMMENTED** [review]
minor questions, looks good overall.

File: src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs (hunk 3 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs]: Per https://msdn.microsoft.com/en-us/library/system.reflection.assembly.getentryassembly(v=vs.110).aspx

```
Can return null when called from unmanaged code.
```

While technically not going to happen in our case, if someone's using our libraries it could break them. 

Additionally, this will change once we have inserted the plugins in VS/dotnet CLI. I will call that out explicitly in the PR.

File: src/NuGet.Clients/NuGet.CommandLine/Commands/Command.cs (hunk 1 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.CommandLine/Commands/Command.cs:191]: This becomes more important later in https://github.com/NuGet/NuGet.Client/pull/2208/files

We only want to discover MSBuild when we absolutely have to. 
That means we only do it when we attempt to use the plugins.

File: src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs (hunk 2 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs:4]: they are sorted? 

Internal namespaces always come after the system namespaces.

File: src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs (hunk 4 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs]: File exists does not account for the casing. 
So it can get weird on Linux based FSs.

Another approach would be to just have an undefined behavior for the case sensitive use-cases. Which might be fine in our case anyways. 

I don't have a hard opinion one way or another, and the only thing required is for me to call out the behavior explicitly in the spec. 

What do you think? @rohit21agrawal

**Comment by nkolev92** [comment]
🔔 @rohit21agrawal 
Answered your questions.

File: src/NuGet.Clients/NuGet.CommandLine/Commands/Command.cs (hunk 1 above)
Comment (rohit21agrawal) [reply src/NuGet.Clients/NuGet.CommandLine/Commands/Command.cs:191]: that's fine. But given that calling this function essentially means that msbuild is going to be discovered for sure, shouldn't it be the responsibility of the caller to do the discovery and pass the path directly to this function?

File: src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs (hunk 4 above)
Comment (rohit21agrawal) [reply src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs]: Directoy.EnumerateFiles can take a major hit on perf, so i would rather ignore cases sensitive use-cases. In any case, the only part where casing could differ is "pluginName.exe/.dll" , which usually won't change casing given how we ship these assemblies.


If perf is not critical here, then i am fine with this.

File: src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+#else
+                "netcore"
+#endif
+                );
+        }
+
+        public static IEnumerable<string> GetConventionBasedPlugins(IEnumerable<string> directories)
+        {
+            var paths = new List<string>();
+            foreach (var directory in directories.Where(Directory.Exists))
```
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs]: how is this list of directories being constructed?

can the responsibility of makign sure that the directory exists be delegated to the caller and you make this function safe by catching the appropriate exception ?

File: src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoverer.cs [hunk 6]
```diff
@@ -178,7 +178,8 @@ private IEnumerable<string> GetPluginFilePaths()
         {
             if (string.IsNullOrEmpty(_rawPluginPaths))
             {
-                return Enumerable.Empty<string>();
+                var directories = new List<string> { PluginDiscoveryUtility.GetNuGetHomePluginsPath(), PluginDiscoveryUtility.GetInternalPlugins() };
```
Comment (jainaashish) [inline src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoverer.cs:181]: `PluginDiscoveryUtility.GetInternalPlugins()` could return nulll which will just be added in `directories` list which will further be processed in next API. You would need to better handle nulls there.

File: src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs [hunk 7]
```diff
@@ -0,0 +1,59 @@
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using NuGet.Common;
+
+namespace NuGet.Protocol.Plugins
+{
+    public static class PluginDiscoveryUtility
+    {
+        public static string InternalPluginDiscoveryRoot { get; set; } = null;
```
Comment (jainaashish) [inline src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs]: wouldn't it be null by default?

File: src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs (hunk 5 above)
Comment (jainaashish) [inline src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs]: `foreach (var directory in directories.Where(Directory.Exists))`
This is really error prone design, since iterating a list and the same time if someone else updates this `directories` list then it will randomly throws `ArgumentException` without much details to debug it. I'd recommend first filtering `directories` list for exists list and then iterate.

Not directly related to this, but similar design issue - https://stackoverflow.com/questions/1167333/why-doesnt-a-foreach-loop-work-in-certain-cases

File: src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs (hunk 4 above)
Comment (jainaashish) [reply src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs]: I agree with Rohit here, `EnumerateFiles` might have a perf hit depending on the plugin size. There must be a better way to check file exists across platform.

File: src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoverer.cs [hunk 8]
```diff
@@ -178,7 +178,8 @@ private IEnumerable<string> GetPluginFilePaths()
...
+                return PluginDiscoveryUtility.GetConventionBasedPlugins(directories);
```
Comment (jainaashish) [inline src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoverer.cs:182]: Just wondering where are we using `NUGET_PLUGIN_PATHS` mechanism to first discover plugin? is that not part of this PR?

File: src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs (hunk 4 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs]: @jainaashish Unfortunately I wasn't able to find a better solution :)

I will rework it to where it doesn't consider casing at all, but it's more performant.

File: src/NuGet.Clients/NuGet.CommandLine/Commands/Command.cs (hunk 1 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.CommandLine/Commands/Command.cs:191]: This call is made for every command. 

MSBuild needs to discovered only when we have restore or we're using http. 
This approach ensures that.

File: src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs (hunk 5 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs]: It's up to the caller to construct the list.
Why shouldn't this function be safe itself?

File: src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoverer.cs (hunk 8 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoverer.cs:182]: rawPluginPaths are the NUGET_PLUGIN_PATHS

File: src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoverer.cs (hunk 6 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoverer.cs:181]: GetConventionBasedPlugins considers nulls already. 
There would be no runtime exceptions currently. 

Additionally, the only time GetInternalPlugins would return a null is if it's called from unmanaged code.

File: src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs (hunk 5 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs]: Decoupled the enumerate call and the for loop.

**Comment by nkolev92** [comment]
🔔 
@jainaashish @rohit21agrawal

**Review by jainaashish — APPROVED** [review]

**Review by rohit21agrawal — APPROVED** [review]
